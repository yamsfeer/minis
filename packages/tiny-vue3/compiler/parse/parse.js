export const nodeTypes = {
  Root: Symbol('Root'),
  Element: Symbol('Element'),
  Attributes: Symbol('Attributes'),
  Interpolation: Symbol('Interpolation'),
  Expression: Symbol('Expression'),
  Text: Symbol('Text'),
}

const State = {
  DATA: Symbol('DATA'),
  RAWTEXT: Symbol('RAWTEXT'),
}

// tokenize 和 analyse 是 parse 的基本实现
// 以下是完整版 parse，使用递归下降算法
export default function parse(str) {
  const context = {
    source: str,
    state: State.DATA,
    consume(num) {
      if (typeof num === 'string') {
        num = num.length
      }
      context.source = context.source.slice(num)
    },
    consumeSpaces() {
      const match = /^[\t\r\n\f\s]+/.exec(context.source)
      match && context.consume(match[0].length)
    }
  }

  const ancestors = []
  const nodes = parseChildren(context, ancestors)

  return {
    type: nodeTypes.Root,
    children: nodes
  }
}

function isEnd(context, ancestors) {
  if (!context.source) {
    return true // source 处理完成
  }

  // 祖先中找到匹配的标签, ancestors 是一个栈
  for (let i = ancestors.length - 1; i >= 0; i--) {
    if (context.source.startsWith(`</${ancestors[i].tag}`)) {
      return true
    }
  }

  return false
}

function parseChildren(context, ancestors) {
  const nodes = []

  while (!isEnd(context, ancestors)) {
    let node
    if (context.state === State.DATA) {
      // 注意这里的 source 都要往 context 取最新值
      if (context.source[0] === '<' && isAlphabet(context.source[1])) {
        node = parseElement(context, ancestors) // 标签元素
      } else if (context.source.startsWith('{{')) {
        node = parseInterpolation(context) // 插值
      }
    }

    if (!node) {
      node = parseText(context)
    }

    nodes.push(node)
  }

  return nodes
}

function isAlphabet(char) {
  return /[a-z]/i.test(char)
}

function parseTag(context, type = 'start') {
  const { consume, consumeSpaces } = context
  // <div name="xxx"></div>
  const startTagReg = /^<([a-z][^[\t\r\n\f\s/>]*)/i
  const endTagReg = /^<\/([a-z][^[\t\r\n\f\s/>]*)/i

  const [match, tag] = type === 'start'
    // match: '<div', tag: 'div'
    ? startTagReg.exec(context.source)
    // match: '</div', tag: 'div'
    : endTagReg.exec(context.source)

  consume(match)
  consumeSpaces()

  // attributes
  const props = parseAttributes(context)

  const isSelfClosing = context.source.startsWith('/>')
  consume(isSelfClosing ? '/>' : '>')

  return {
    type: nodeTypes.Element,
    tag,
    props,
    children: [],
    isSelfClosing
  }
}

function parseAttributes(context) {
  const { consume, consumeSpaces } = context
  const props = []

  while (
    !context.source.startsWith('>') &&
    !context.source.startsWith('/>')
  ) {
    const [name] = /^[^\t\r\n\f\s/>][^\t\r\n\f\s/>=]*/.exec(context.source)

    consume(name)
    consumeSpaces()
    consume('=')
    consumeSpaces()

    let value = ''
    const DOUBLE_QUOTE = `"`
    const SINGLE_QUOTE = `'`
    const [quote] = context.source
    const isQuoted = quote === SINGLE_QUOTE || quote === DOUBLE_QUOTE

    if (isQuoted) {
      consume(quote)
      const endQuoteIndex = context.source.indexOf(quote)
      if (endQuoteIndex > -1) {
        value = context.source.slice(0, endQuoteIndex)
        consume(value)
        consume(quote)
      } else {
        console.error('引号不匹配')
      }
    } else {
      [value] = /^[^\t\r\n\f\s>]+/.exec(context.source)
      consume(value)
    }
    consumeSpaces()

    props.push({
      type: nodeTypes.Attributes,
      name,
      value
    })
  }

  return props
}

function parseElement(context, ancestors) {
  const element = parseTag(context)
  if (element.isSelfClosing)
    return element

  ancestors.push(element)
  // 每次 parseChildren 之前的空格换行怎么处理
  context.consumeSpaces()
  element.children = parseChildren(context, ancestors)
  context.consumeSpaces()
  ancestors.pop()

  if (context.source.startsWith(`</${element.tag}`)) {
    parseTag(context, 'end')
  } else {
    console.error(`${element.tag}缺少闭合标签`)
  }

  return element
}

function parseText(context) {
  let index = context.source.length // 有可能后面全是文本
  const quoteIndex = context.source.indexOf('<')
  const delimiterIndex = context.source.indexOf('{{')

  // 提取 '<' 之前的内容
  if (quoteIndex > -1) {
    index = quoteIndex
  }
  // 在 '<' 之前还有 '{{'，说明是文本中的插值，提取 '{{' 之前的内容
  if (delimiterIndex > -1 && delimiterIndex < index) {
    index = delimiterIndex
  }

  const content = context.source.slice(0, index)
  context.consume(content)

  return {
    type: nodeTypes.Text,
    content
  }
}

function parseInterpolation(context) {
  const { consume } = context

  consume('{{')
  const closeIndex = context.source.indexOf('}}')
  if (closeIndex < 0) {
    console.error('缺少结束定界符')
  }

  const content = context.source.slice(0, closeIndex)

  consume(content)
  consume('}}')

  return {
    type: nodeTypes.Interpolation,
    content: {
      type: nodeTypes.Expression,
      content
    }
  }
}
