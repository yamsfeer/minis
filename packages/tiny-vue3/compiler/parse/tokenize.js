const State = {
  data: Symbol('data'),

  tagOpen: Symbol('tagOpen'),
  tagName: Symbol('tagName'),

  text: Symbol('text'),

  tagEnd: Symbol('tagEnd'),
  tagEndName: Symbol('tagEndName'),
}

export const tokenTypes = {
  tagOpen: Symbol('tagOpen'),
  text: Symbol('text'),
  tagEnd: Symbol('tagEnd'),
}

const createTagOpen = (tagName) => ({ type: tokenTypes.tagOpen, name: tagName })
const createText = (text) => ({ type: tokenTypes.text, content: text })
const createTagEnd = (tagName) => ({ type: tokenTypes.tagEnd, name: tagName })

export default function tokenize(str) {
  const tokens = []
  let state = State.data
  let tmp = '' // 字符拼接
  let c

  while (str.length) {
    c = str[0] // 当前字符
    switch (state) {
      case State.data:
        if (c === '<') { state = State.tagOpen }
        else if (isAlphabet(c)) {
          state = State.text
          tmp += c
        }

        break
      case State.tagOpen:
        if (isAlphabet(c)) {
          state = State.tagName
          tmp += c
        }
        else if (c === '/') { state = State.tagEnd }

        break
      case State.tagName:
        if (isAlphabet(c)) { tmp += c }
        else if (c === '>') { // 取得一个 tagName
          state = State.data
          tokens.push(createTagOpen(tmp))
          tmp = ''
        }

        break
      case State.text:
        if (isAlphabet(c)) { tmp += c }
        else if (c === '<') {
          state = State.tagOpen
          tokens.push(createText(tmp))
          tmp = ''
        }

        break
      case State.tagEnd:
        if (isAlphabet(c)) {
          state = State.tagEndName
          tmp += c
        }
        break
      case State.tagEndName:
        if (isAlphabet(c)) { tmp += c }
        else if (c === '>') {
          state = State.data
          tokens.push(createTagEnd(tmp))
          tmp = ''
        }
        break

      default: break
    }
    str = str.slice(1) // consume
  }

  return tokens
}

function isAlphabet(char) {
  return /[a-zA-Z0-9]/.test(char)
}
