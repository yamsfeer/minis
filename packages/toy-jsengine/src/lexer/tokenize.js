import {
  createIdentifierOrKeyword,
  createNumber,
  createOperator,
  createString,
  createBracket,
} from './token.js'

import {
  isSpace,
  isQuote,
  isLiteral,
  isNumber,
  isStrQuote,
  isOperator,
  createPeekIterator,
} from '../utils/index.js'

export const tokenize = (code) => {
  const it = createPeekIterator(code)
  const tokens = []

  while (it.hasNext()) {
    const c = it.peek()

    // 空格，换行符，缩进等
    if (isSpace(c)) {
      it.next()
      continue
    }

    // 括号
    if (isQuote(c)) {
      tokens.push(createBracket(it))
      continue
    }

    // 标识符或关键字，不能以数字开头
    if (isLiteral(c) && !isNumber(c)) {
      tokens.push(createIdentifierOrKeyword(it))
      continue
    }

    // number
    if (isNumber(c)) {
      tokens.push(createNumber(it))
      continue
    }

    // 字符串，单引号，双引号，反引号
    if (isStrQuote(c)) {
      tokens.push(createString(it))
      continue
    }

    // 操作符
    if (isOperator(c)) {
      tokens.push(createOperator(it))
      continue
    }

    // 最后剩下一个 undefined
    if (c === undefined) {
      break
    }

    throw lexicalException(`Unexpected character ${c}`)
  }

  return tokens
}

export const lexicalException = (msg) => new Error(msg)
