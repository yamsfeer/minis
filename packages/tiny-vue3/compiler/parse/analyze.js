import { tokenTypes } from './tokenize.js'

export const nodeTypes = {
  Root: Symbol('Root'),
  Element: Symbol('Element'),
  Text: Symbol('Text'),
}

const createRootNode = () => ({
  type: nodeTypes.Root,
  children: []
})

const createElementNode = (name) => ({
  type: nodeTypes.Element,
  tag: name,
  children: []
})

const createTextNode = (content) => ({
  type: nodeTypes.Text,
  content
})


export default function analyse(tokens) {
  const root = createRootNode()
  const stack = [root]

  while (tokens.length) {
    const parent = stack[stack.length - 1]
    const token = tokens.shift()

    switch (token.type) {
      case tokenTypes.tagOpen:
        const elementNode = createElementNode(token.name)

        parent.children.push(elementNode)
        stack.push(elementNode)
        break
      case tokenTypes.text:
        const textNode = createTextNode(token.content)
        parent.children.push(textNode)
        break
      case tokenTypes.tagEnd:
        stack.pop()
        break
    }
  }

  return root
}
