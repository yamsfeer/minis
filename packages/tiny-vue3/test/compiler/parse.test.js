import tokenize from '../../compiler/parse/tokenize.js'
import analyse from '../../compiler/parse/analyze.js'
import parse from '../../compiler/parse/parse.js'

describe('compiler:basic parse', () => {
  it('basic tokenize for tag', () => {
    const template = `<div><p>hello</p></div>`
    const tokens = tokenize(template)

    expect(tokens).toMatchSnapshot()
  })
  it('basic analyse', () => {
    const template = `<div>
      <p>hello</p>
      <p>world</p>
    </div>`
    const tokens = tokenize(template)
    const ast = analyse(tokens)

    expect(ast).toMatchSnapshot()
  })
})

describe('compiler:parse', () => {
  it('complete parse', () => {
    // 一开始的空格换行？？
    // 没有 value 的 attribute ？？ disabled 出错
    // 每次 parseChildren 之前的空格换行怎么处理
    const template = `<div foo="foo" bar='bar' baz=foo @click="handleClick">
        <p :title="title" v-show="isShow">
          hello {{ data }}
        </p>
      </div>
    `
    const ast = parse(template)
    expect(ast).toMatchSnapshot()
  })
})
