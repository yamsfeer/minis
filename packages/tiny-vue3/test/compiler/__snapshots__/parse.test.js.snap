// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler:basic parse > basic analyse 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "hello",
              "type": Symbol(Text),
            },
          ],
          "tag": "p",
          "type": Symbol(Element),
        },
        {
          "children": [
            {
              "content": "world",
              "type": Symbol(Text),
            },
          ],
          "tag": "p",
          "type": Symbol(Element),
        },
      ],
      "tag": "div",
      "type": Symbol(Element),
    },
  ],
  "type": Symbol(Root),
}
`;

exports[`compiler:basic parse > basic tokenize for tag 1`] = `
[
  {
    "name": "div",
    "type": Symbol(tagOpen),
  },
  {
    "name": "p",
    "type": Symbol(tagOpen),
  },
  {
    "content": "hello",
    "type": Symbol(text),
  },
  {
    "name": "p",
    "type": Symbol(tagEnd),
  },
  {
    "name": "div",
    "type": Symbol(tagEnd),
  },
]
`;

exports[`compiler:parse > complete parse 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "hello ",
              "type": Symbol(Text),
            },
            {
              "content": {
                "content": " data ",
                "type": Symbol(Expression),
              },
              "type": Symbol(Interpolation),
            },
            {
              "content": "
        ",
              "type": Symbol(Text),
            },
          ],
          "isSelfClosing": false,
          "props": [
            {
              "name": ":title",
              "type": Symbol(Attributes),
              "value": "title",
            },
            {
              "name": "v-show",
              "type": Symbol(Attributes),
              "value": "isShow",
            },
          ],
          "tag": "p",
          "type": Symbol(Element),
        },
        {
          "content": "
      ",
          "type": Symbol(Text),
        },
      ],
      "isSelfClosing": false,
      "props": [
        {
          "name": "foo",
          "type": Symbol(Attributes),
          "value": "foo",
        },
        {
          "name": "bar",
          "type": Symbol(Attributes),
          "value": "bar",
        },
        {
          "name": "baz",
          "type": Symbol(Attributes),
          "value": "foo",
        },
        {
          "name": "@click",
          "type": Symbol(Attributes),
          "value": "handleClick",
        },
      ],
      "tag": "div",
      "type": Symbol(Element),
    },
    {
      "content": "
    ",
      "type": Symbol(Text),
    },
  ],
  "type": Symbol(Root),
}
`;

exports[`compiler:basic parse basic tokenize for tag 1`] = `
[
  {
    "name": "div",
    "type": Symbol(tagOpen),
  },
  {
    "name": "p",
    "type": Symbol(tagOpen),
  },
  {
    "content": "hello",
    "type": Symbol(text),
  },
  {
    "name": "p",
    "type": Symbol(tagEnd),
  },
  {
    "name": "div",
    "type": Symbol(tagEnd),
  },
]
`;

exports[`compiler:basic parse basic analyse 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "hello",
              "type": Symbol(Text),
            },
          ],
          "tag": "p",
          "type": Symbol(Element),
        },
        {
          "children": [
            {
              "content": "world",
              "type": Symbol(Text),
            },
          ],
          "tag": "p",
          "type": Symbol(Element),
        },
      ],
      "tag": "div",
      "type": Symbol(Element),
    },
  ],
  "type": Symbol(Root),
}
`;

exports[`compiler:parse complete parse 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "hello ",
              "type": Symbol(Text),
            },
            {
              "content": {
                "content": " data ",
                "type": Symbol(Expression),
              },
              "type": Symbol(Interpolation),
            },
            {
              "content": 
"
        "
,
              "type": Symbol(Text),
            },
          ],
          "isSelfClosing": false,
          "props": [
            {
              "name": ":title",
              "type": Symbol(Attributes),
              "value": "title",
            },
            {
              "name": "v-show",
              "type": Symbol(Attributes),
              "value": "isShow",
            },
          ],
          "tag": "p",
          "type": Symbol(Element),
        },
        {
          "content": 
"
      "
,
          "type": Symbol(Text),
        },
      ],
      "isSelfClosing": false,
      "props": [
        {
          "name": "foo",
          "type": Symbol(Attributes),
          "value": "foo",
        },
        {
          "name": "bar",
          "type": Symbol(Attributes),
          "value": "bar",
        },
        {
          "name": "baz",
          "type": Symbol(Attributes),
          "value": "foo",
        },
        {
          "name": "@click",
          "type": Symbol(Attributes),
          "value": "handleClick",
        },
      ],
      "tag": "div",
      "type": Symbol(Element),
    },
    {
      "content": 
"
    "
,
      "type": Symbol(Text),
    },
  ],
  "type": Symbol(Root),
}
`;
